src/client.rs

```diff
@@ -88,7 +153,106 @@ impl Record {
+
+            match maybe_attachment {
+                Some(Ok(meta)) => {
+                    self.attachment = Attachment::Some(meta);
+                }
+                Some(Err(err)) => {
+                    // serde_json::Error does not implement std::error::Error,
+                    // weirdly, so convert it to a string.
+                    return Some(Err(ClientError::AttachmentFormatError(
+                        anyhow!("Could not convert attachment to requested type: {}", err)
+                    )));
+                }
+                None => {
+                    self.attachment = Attachment::None;
+                }
+            }
+        }
+
+        match self.attachment {
+            Attachment::Some(ref metadata) => {
+                Some(Self::fetch_attachment::<T, E>(client, metadata).await)
+            },
+            Attachment::None => Option::None,
+            Attachment::Pending => unreachable!(),
+        }
+    }
+
+    #[cfg(feature = "attachments")]
+    async fn fetch_attachment<T, E>(
+        client: &mut Client,
+        metadata: &AttachmentMetadata
+    ) -> Result<T, ClientError>
+    where T: TryFrom<Vec<u8>, Error = E>,
+          E: 'static + Send + Sync + std::error::Error
+    {
+        let key = format!("attachment:{}", metadata.hash);
+        let bytes = match client.storage.retrieve(&key) {
+            Ok(bytes) => Ok(bytes),
+
+            Err(StorageError::KeyNotFound { .. }) => {
+                let url = {
+                    let server_info = client.server_info().await?;
+                    match &server_info["capabilities"]["attachments"]["base_url"] {
+                        Value::String(s) => Url::parse(s).map_err(|err|
+                            ClientError::APIError(KintoError::URLError(err))
+                        )?,
+                        Value::Null => {
+                            return Err(ClientError::CompatibilityError(
+                                anyhow!("server does not support attachments")
+                            ));
+                        }
+                        _ => {
+                            return Err(ClientError::CompatibilityError(
+                                anyhow!("server did not return a valid attachment base_url")
+                            ));
+                        },
+                    }
+                };
+
+                let response = client.http_client.get(url).await
+                    .map_err(|_| ClientError::APIError(KintoError::HTTPBackendError()))?;
+
+                if response.is_success() {
+                    Ok(response.body)
+                } else {
+                    return Err(ClientError::APIError(KintoError::UnexpectedResponse {
+                        url: client.server_url.clone(),
+                        response,
+                    }));
+                }
+            }
+
+            Err(err) => Err(ClientError::StorageError(err)),
+        }?;
```

Comment:
Note for later: you would have to check the size / SHA256 / here (see https://searchfox.org/mozilla-central/rev/9a5f36b0ddb9cb8ae556fc5b45f8ccea0f0da6f8/services/settings/SharedUtils.jsm#24-36)

Reply:
Ah yes! I meant to ask: Is there a preferred way to calculate that hash already in RS? If not I can pull in something from crates.io easily enough.

Reply:
After implementing this, I think we shouldn't verify size, only the hash. Attachments of different sizes will (be likely to) have different hashes, and Merino wants to be able to make all content checks optional. The alternative would be something like having the verifier check the size, or just check the entire content. Since checking size along side signatures is such a weak check, I don't think that's worth the extra complexity.

Reply:
For us it was only a way to bypass hash computation if sizes differ.